ruth/data/hdf5_writer.py

```diff
@@ -0,0 +1,206 @@
+import re
+import h5py
+import numpy as np
+from collections import OrderedDict
+from datetime import datetime
+
+INT32_MAX = 2147483647
+
+
+def coord_to_int(coord):
+    return int(round(coord * pow(10, 6)))
+
+
+def get_osmid_from_data(data):
+    return data['osmid'][0] if type(data['osmid']) is list else data['osmid']
+
+
+def get_speed_from_data(data):
+    if 'speed_kph' in data:
+        return data['speed_kph']
+
+    speed_str = data.get('maxspeed', "50")
+    speed_str = speed_str[0] if type(speed_str) is list else speed_str
+
+    nums = re.findall(r'\d+', speed_str)
+    r = int(nums[0]) if len(nums) > 0 else 50
+    return r
+
+
+def dict_to_dataset(dictionary, dtype):
+    data_array = []
+
+    for _, value in dictionary.items():
+        data_array.append([value])
+
+    return np.array(data_array, dtype=dtype)
+
+
+def save_graph_to_hdf5(g, file_path):
+    part_name = b'CZE'
+    edge_index = 0
+    node_type = [
+        ("id", np.int32),
+        ("latitudeInt", np.int32),
+        ("longitudeInt", np.int32),
+        ("edgeOutCount", np.uint8),
+        ("edgeOutIndex", np.int32),
+        ("edgeInCount", np.uint8)
+    ]
+    edge_type = [
+        ("edgeId", np.int32),
+        ("nodeIndex", np.int32),
+        ("computed_speed", np.int32),
+        ("length", np.int32),
+        ("edgeDataIndex", np.int32)
+    ]
+    edge_data_type = [
+        ('id', np.int32),
+        ('speed', np.uint8),
+        ('funcClass', np.uint8),
+        ('lanes', np.uint8),
+        ('vehicleAccess', np.uint8),
+        ('specificInfo', np.uint16),
+        ('maxWeight', np.uint16),
+        ('maxHeight', np.uint16),
+        ('maxAxleLoad', np.uint8),
+        ('maxWidth', np.uint8),
+        ('maxLength', np.uint8),
+        ('incline', np.int8)
+    ]
+    node_map_type = [
+        ('nodeId', np.int32),
+        ('partId', np.dtype('S4')),
+        ('nodeIndex', np.int32)
+    ]
+
+    nodes_array = []
+    edges_array = []
+    node_dict = OrderedDict()
+    edge_data_dict = OrderedDict()
+
+    ids_from_64_to_32 = {}
```

Comment:
This maps OSM IDs into HDF5 IDs, so I think a better name would be something like `osm_to_hdf_map` or something like that.

Why do we even need the mapping? Why can't we use the OSM IDs directly in the HDF5 file, when we generate it from scratch?

Reply:
Because the OSM IDs are bigger than max int32 max value, which is used to store the IDs in the c++ code.

Reply:
I see. Then please add a comment that explains this to this function. It could be nice to find the minimum OSM ID and subtract it from all the IDs, so that the "gaps" between the IDs stay the same, for simple identification and debugging. But it wouldn't work if max - min ID is still larger than 2^32. Let's just create fresh IDs for now, as you already do.

One additional question: what's the reason of the `+ 1000` addition?

Reply:
The node IDs are for our current map are in range from 172 512 to 11 068 325 603, so I think the substraction wouldn't be possible.
With the indexing I think we had some issues indexing from 0. And then indexing from 1 was confusing for debugging because node 1 had a node_map index of 0 etc. So we randomly selected +1000 to have the last digits the same for node IDs and node_map indexes. Should I change it to +1 or +10 maybe?

Reply:
1000 is fine, just again please include a comment that explains why this was done, otherwise it's not clear why is it there :)

Reply:
I added the explanatory comments.